x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/cluster/metadata/MetadataMigrateToDataTiersRoutingService.java

```diff
@@ -0,0 +1,309 @@
+            if (allocateActionDefinesRoutingRules(nodeAttrName, allocateAction)) {
+                Map<String, LifecycleAction> actionMap = new HashMap<>(phase.getActions());
+                // this phase contains an allocate action that defines a require rule for the attribute name so we'll remove all the
+                // rules to allow for the migrate action to be injected
+                if (allocateAction.getNumberOfReplicas() != null) {
+                    // keep the number of replicas configuration
+                    AllocateAction updatedAllocateAction =
+                        new AllocateAction(allocateAction.getNumberOfReplicas(), null, null, null);
+                    actionMap.put(allocateAction.getWriteableName(), updatedAllocateAction);
+                    logger.debug("ILM policy [{}], phase [{}]: updated the allocate action to [{}]", lifecyclePolicy.getName(),
+                        phase.getName(), allocateAction);
+                } else {
+                    // remove the action altogether
+                    actionMap.remove(allocateAction.getWriteableName());
+                    logger.debug("ILM policy [{}], phase [{}]: removed the allocate action", lifecyclePolicy.getName(),
+                        phase.getName());
+                }
+
+                Phase updatedPhase = new Phase(phase.getName(), phase.getMinimumAge(), actionMap);
+                Map<String, Phase> updatedPhases =
+                    new HashMap<>(newLifecyclePolicy == null ? lifecyclePolicy.getPhases() : newLifecyclePolicy.getPhases());
+                updatedPhases.put(phaseEntry.getKey(), updatedPhase);
+                newLifecyclePolicy = new LifecyclePolicy(lifecyclePolicy.getName(), updatedPhases);
+            }
+        }
+        return newLifecyclePolicy;
+    }
+
+    static boolean allocateActionDefinesRoutingRules(String nodeAttrName, @Nullable AllocateAction allocateAction) {
+        return allocateAction != null && (allocateAction.getRequire().get(nodeAttrName) != null ||
+            allocateAction.getInclude().get(nodeAttrName) != null ||
+            allocateAction.getExclude().get(nodeAttrName) != null);
+    }
+
+    static List<String> migrateIndices(Metadata.Builder mb, ClusterState currentState, String nodeAttrName) {
+        List<String> migratedIndices = new ArrayList<>();
+        String nodeAttrIndexRequireRoutingSetting = INDEX_ROUTING_REQUIRE_GROUP_SETTING.getKey() + nodeAttrName;
+        String nodeAttrIndexIncludeRoutingSetting = INDEX_ROUTING_INCLUDE_GROUP_SETTING.getKey() + nodeAttrName;
+        for (ObjectObjectCursor<String, IndexMetadata> index : currentState.metadata().indices()) {
+            IndexMetadata indexMetadata = index.value;
+            Settings currentSettings = indexMetadata.getSettings();
+            Settings newSettings = migrateIndexSettings(nodeAttrIndexRequireRoutingSetting, indexMetadata);
+            if (newSettings.equals(currentSettings)) {
```

Comment:
I *think* because of this change (if I read this correctly), if you have both a require and include, only the `require` will be removed, so you'd end up with a before like:

```
{
  "require": "hot",
  "include": "hot"
}
```

And an after like:

```
{
  "_tier_preference": "data_hot",
  "include": "hot"
}
```

I think maybe `migrateIndexSettings` should remove all three types (require, include, exclude) whenever it encounters the attribute passed in and sets the new tier preference, or else remove them if in this method if the settings have changed and they have the right value?

Reply:
Yes, so I chose to say "we migrate `require.data` **or** if that's not present look at trying to migrate `include.data`" (which I think would cover most cases I've seen in the wild. Also all our `hot-warm` architecture docs talk about the `require` setting being used)

> should remove all three types (require, include, exclude) whenever it encounters the attribute passed in and sets the new tier preference, or else remove them if in this method if the settings have changed and they have the right value?

I believe this could be a good compromise (that might possibly erase some configurations that attempt to implement sub-tiering using the same attribute name, but that's an edge case configuration I reckon). 

I'll change the implementation to still **firstly** look at `require.data` and if we migrate this setting remove the `include.data` and `exclude.data` altogether. 
If `require.data` is not present (or not migrateable - by virtue of using "the_hot_node" as value) we look at `include.data` and if migrating of `include` is successful we'll remove the `require.data` and `exclude.data` settings.
So both
```
{
  "require": "hot",
  "include": "hot"
}
```
and
```
{
  "require": "the_hot_node",
  "include": "hot"
}
```
will be migrated to 
```
{
  "_tier_preference": "data_hot"
}
```

Reply:
That sounds good to me, thanks for changing this